[Model Monitoring] Fix avoid `get_model_endpoint` call from serving

royischoss opened the pull request:
https://iguazio.atlassian.net/browse/ML-9265
Changes: 
1) `V2ModelServer` and `VotingEnsemble` avoid multiple API calls from `_lazy_init`, using class args for background task state, and model endpoints needed fields.
2) Change in the behavior of `create_model_endpoints` background task, divided to two parts: instructions, and the actual creation - allowing to pass mep uids to the serving classes
3) small changes in the instructions creation picking uid early for the mep while respecting creation_strategy
4) Checking background task status in   `BaseModelRouter::_update_background_task_state` handling every 15 seconds avoiding multiple api calls for each invoke

royischoss committed:
1. Divie `create_model_endpoints` into the actual creation and to the instruction creation - allows to pass class arguments to function
2. chagne `V2ModelServer:lazy_init` so it won't use `get_model_endpoint` crud call and will use the class arguments or the model artifact spec
3. Taking out the background task check to `ModelRouter:_handle_event` every 15 seconds till the task of mep creation is over

royischoss committed:
monitoring works well with some changes need to be made in `_fill_model_endpoint_with_model_data`

royischoss committed:
Merge remote-tracking branch 'origin/development' into ml-9265

# Conflicts:
#	server/py/framework/db/sqldb/db.py

royischoss committed:
adding after merge change in list_model_endpoints, routes.py adding the background task check to VotingEnsemble and ParallelRun

royischoss committed:
lint

royischoss committed:
lint

royischoss committed:
lint and ci

royischoss committed:
fix ci and lint

royischoss committed:
another fix

royischoss committed:
Merge remote-tracking branch 'origin/development' into ml-9265

royischoss committed:
log fix and bug fix and removed TODO

royischoss committed:
add termination flag to mock not wasting calls to background check for each route

royischoss committed:
Merge remote-tracking branch 'origin/development' into ml-9265

royischoss committed:
lint

davesh0812 commented on mlrun/serving/routers.py:
```diff
@@ -195,6 +258,8 @@ def _resolve_route(self, body, urlpath):
 
     def _handle_event(self, event):
         name, route, subpath = self._resolve_route(event.body, event.path)
+        self._update_background_task_state(event)
```
I think all the the router uses `_pre_handle_event` so lets put `_update_background_task_state` only in there.. what do you think?

davesh0812 commented on mlrun/serving/routers.py:
```diff
@@ -160,6 +164,65 @@ def postprocess(self, event):
+                        f"Model endpoint creation task is still in progress with the current state: "
+                        f"{background_task.status.state}. This event will not be monitored.",
+                        name=self.name,
+                        event_id=event_id,
```
lets remove this and add maybe timestamp that explain that all the event will not be monitoring from now to now+15sec

davesh0812 commented on server/py/framework/db/sqldb/db.py:
```diff
@@ -7563,8 +7563,12 @@ def list_model_endpoints(
         order_by: typing.Optional[str] = None,
-    ) -> mlrun.common.schemas.ModelEndpointList:
-        model_endpoints: list[mlrun.common.schemas.ModelEndpoint] = []
+        as_dict: bool = False,
+    ) -> Union[mlrun.common.schemas.ModelEndpointList, list[ModelEndpoint]]:
```
Your is call `as_dict` but you returning list?

davesh0812 commented on server/py/services/api/api/endpoints/nuclio.py:
```diff
@@ -271,19 +271,30 @@ async def deploy_function(
         background_tasks=[]
     )
     if function.get("kind") == mlrun.runtimes.RuntimeKinds.serving:
+        db_session = framework.db.session.create_session()
```
You can use here the `db_session` you got as param

davesh0812 commented on server/py/services/api/crud/model_monitoring/deployment.py:
```diff
@@ -1440,13 +1486,22 @@ def _extract_model_endpoints_from_function_graph(
+            mlrun.serving.states.RouterStep, mlrun.serving.states.RootFlowStep
+        ],
     ]:
+        endpoints_names_uid_map = {
+            endpoint.model_name: endpoint.uid for endpoint in model_endpoints
+        }
```
You should do as part of the list db function. Only in there the key needs to be - "<project_name>-<function_name>-<function_tag>-<name>".

davesh0812 reviewed (changes requested):
Nice. Few comments :)

royischoss committed:
CR comments fix -
1) change list_model_endpoints to return a dictionary while using as_dict
2) change the location of background task check
3) change the creation and clossing of db session in nuclio.py

royischoss committed:
fixes for mep creation uid

royischoss committed:
fix remove filter string

royischoss committed:
Merge remote-tracking branch 'origin/development' into ml-9265

royischoss committed:
fix naming for model_endpoints_dict

royischoss committed:
Merge remote-tracking branch 'origin/development' into ml-9265

royischoss committed:
fixes cr

royischoss committed:
Merge remote-tracking branch 'origin/development' into ml-9265
